File: scripts/pa-annotate.pl
#!/usr/bin/env perl

# Need a license here

use strict;
use warnings;
use Getopt::Long::Descriptive;
use Bio::KBase::probabilistic_annotation::Client;
use Bio::KBase::probabilistic_annotation::Helpers qw(get_probanno_client);
use Bio::KBase::workspaceService::Helpers qw(auth workspace printObjectMeta);

my $manpage =
"
NAME
      pa-annotate -- generate probabilistic annotation for a genome

SYNOPSIS
      pa-annotate <Genome ID> <ProbAnno ID> [OPTIONS]

DESCRIPTION
      Generate a probabilistic annotation for a genome.
      
      Options:
      -e, --showerror    Show any errors in execution
      --genomews         ID of workspace where Genome object is stored
      -h, --help         Display this help message, ignore all arguments
      -o, --overwrite    Overwrite existing ProbAnno object with same name
      -v, --verbose      Print verbose messages
      -w, --probannows   ID of workspace where ProbAnno object is saved

EXAMPLES
      Annotate:
      > pa-annotate kb\|g.0 kb.g.0
      
AUTHORS
      Matt Benedict, Mike Mundy
";

# Define usage and options.
my $primaryArgs = [ "Genome ID", "ProbAnno ID" ];
my ( $opt, $usage ) = describe_options(
    'pa-annotate <' . join( "> <", @{$primaryArgs} ) . '> %o',
    [ 'probannows:s', 'ID of workspace where ProbAnno object is saved', { "default" => workspace() } ],
    [ 'genomews:s', 'ID of workspace where Genome object is stored', { "default" => "KBaseCDMGenomes" } ],
    [ 'overwrite:o', "Set as 1 to overwrite existing ProbAnno object with same name", { "default" => 0 } ],
    [ 'showerror|e', 'Set as 1 to show any errors in execution', { "default" => 0 } ],
    [ 'verbose|v', 'Set as 1 to print verbose messages', { "default" => 0 } ],
    [ 'help|h', 'Show help text' ],
    [ 'usage|?', 'Show usage information' ]
    );
if ( defined( $opt->{help} ) ) {
    print $manpage;
    exit 0;
}
if (defined($opt->{usage})) {
	print $usage;
	exit 0;
}

# Process primary arguments.
foreach my $arg ( @{$primaryArgs} ) {
    $opt->{$arg} = shift @ARGV;
    if ( !defined( $opt->{$arg} ) ) {
		print $usage;
		exit;
    }
}

# Create a client object.
my $client = get_probanno_client();

# Define translation from options to function parameters.
my $translation = {
    "Genome ID"   => "genome",
    "ProbAnno ID" => "probanno",
    genomews      => "genome_workspace",
    probannows    => "probanno_workspace",
    overwrite     => "overwrite"
};

# Instantiate parameters for function.
my $params = { auth => auth(), };
foreach my $key ( keys( %{$translation} ) ) {
    if ( defined( $opt->{$key} ) ) {
		$params->{ $translation->{$key} } = $opt->{$key};
    }
}
print "overwrite=".$opt->{overwrite}."\n";

# Call the function.
my $output = $client->annotation_probabilities_id($params);
if (!defined($output)) {
	print "Probabilistic annotation failed!\n";
	exit 1;
} else {
	print "Probabilistic annotation successfully generated in workspace:\n";
	printObjectMeta($output);
}
exit 0;

File: scripts/pa-annotate.py
import argparse
import traceback
import sys
from biokbase.probabilistic_annotation.Helpers import get_url
from biokbase.probabilistic_annotation.Client import ProbabilisticAnnotation
from biokbase.workspace.ScriptHelpers import user_workspace

desc1 = '''
NAME
      pa-annotate -- generate probabilistic annotation for a genome

SYNOPSIS      
'''

desc2 = '''
DESCRIPTION
      Generate alternative annotations for every gene in a genome together
      with their likelihoods.  The current method for calculating likelihoods
      is based on similarity (BLAST) to genes in subsystems and genes with
      literature evidence.
      
      This command takes a significant amount of time to run (since it has to
      run BLAST against a large database), so it is placed on a queue and 
      returns a job ID.  Use the pa-checkjob command to see if your job has
      finished.  When it is done the results are saved in a ProbAnno typed
      object with the specified ID.
      
      The ProbAnno object can be used as input to gapfilling a metabolic model
      using the --probanno option for the fba-gapfill command.
'''

desc3 = '''
EXAMPLES
      Generate probabilistic annotation for E. coli K12 genome:
      > pa-annotate kb|g.0.genome kb|g.0.probanno

SEE ALSO
      pa-calculate
      pa-checkjob
      pa-getprobanno
      fba-gapfill

AUTHORS
      Matt Benedict, Mike Mundy 
'''

if __name__ == "__main__":
    # Parse options.
    parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter, prog='pa-annotate', epilog=desc3)
    parser.add_argument('genome', help='ID of Genome object', action='store', default=None)
    parser.add_argument('probanno', help='ID of ProbAnno object', action='store', default=None)
    parser.add_argument('-w', '--probannows', help='workspace where ProbAnno object is saved', action='store', dest='probannows', default=None)
    parser.add_argument('--genomews', help='workspace where Genome object is saved', action='store', dest='genomews', default=None)
    parser.add_argument('--url', help='url for service', action='store', dest='url', default=None)
    parser.add_argument('-e', '--show-error', help='show detailed information for an exception', action='store_true', dest='showError', default=False)
    usage = parser.format_usage()
    parser.description = desc1 + '      ' + usage + desc2
    parser.usage = argparse.SUPPRESS
    args = parser.parse_args()
    
    # Create input parameters for annotate() function.
    input = dict()
    input['genome'] = args.genome
    input['probanno'] = args.probanno
    if args.genomews is None:
        input['genome_workspace'] = user_workspace()
    else:
        input['genome_workspace'] = args.genomews
    if args.probannows is None:
        input['probanno_workspace'] = user_workspace()
    else:
        input['probanno_workspace'] = args.probannows
                
    # Create a probabilistic annotation client.
    if args.url is None:
        args.url = get_url()
    paClient = ProbabilisticAnnotation(url=args.url)

    # Submit a job to annotate the specified genome.
    try:
        jobid = paClient.annotate(input)
        print 'Probabilistic annotation job '+jobid+' successfully submitted'
    except Exception as e:
        print 'Error starting job: %s' %(e.message)
        if args.showError:
            traceback.print_exc(file=sys.stdout)
        exit(1)

    exit(0)
